File: backend/server/models/userModel.ts
import { Schema, model } from "mongoose";

// define enum for gender and role so no "wrong" values can be used
enum gender {
    male = "male",
    female = "female",
    other = "other"
}

enum role {
    user = "user",
    admin = "admin"

}

// defines the schema for the user collection, this is for the typescript side
export interface UserInterface {
    firstName: string;
    lastName: string;
    birthDate: String;
    email: string;
    gender: gender;
    password: string;
    experience: number;
    role: role;
    avatar?: string;
}

// defines the schema for the user collection, this is how a user should look like in the database
const userSchema = new Schema<UserInterface>({
    firstName: { type: String, required: true },
    lastName: { type: String, required: true },
    birthDate: { type: Date, required: true },
    email: { type: String, required: true },
    gender: { type: String, enum: ["male", "female", "other"], required: true },
    password: { type: String, required: true },
    experience: { type: Number, required: true },
    role: { type: String, enum: ["user", "admin"], required: true },
    avatar: { type: String },
});

export const UserModel = model('User', userSchema);

File: lib/models/user.model.dart
class UserModel {
  const UserModel(
      {required this.firstName,
      required this.lastName,
      required this.email,
      required this.birthDate,
      required this.gender,
      required this.password,
      required this.experience,
      required this.role,
      this.avatar});

  final String firstName;
  final String lastName;
  final String email;
  final DateTime birthDate;
  final String gender;
  final String password;
  final int experience;
  final String role;
  final String? avatar;

  UserModel.fromJson(Map<String, dynamic> json)
      : firstName = json['firstName'] as String,
        lastName = json['lastName'] as String,
        email = json['email'] as String,
        birthDate = DateTime.parse(json['birthDate'] as String),
        gender = json['gender'] as String,
        password = json['password'] as String,
        experience = json['experience'] as int,
        role = json['role'] as String,
        avatar = json['avatar'] as String?;

  Map<String, dynamic> toJson() => {
        'firstName': firstName,
        'lastName': lastName,
        'email': email,
        'birthDate': birthDate,
        'gender': gender,
        'password': password,
        'experience': experience,
        'role': role,
        if (avatar != null) 'avatar': avatar,
      };

  int get level => (experience ~/ 4000);
  double get progress => (experience % 4000) / 4000; // between 0 and 1
}
